### DISCUSSION TITLE
how to make sphinx.ext.todo render in a docstring with sphinx.ext.napoleon

### DISCUSSION THREAD

**Discussion by raybellwaves** [sphinx-doc/sphinx#12306] [Q&A] [answered]
Hi,

I just build the todo extension (https://github.com/raybellwaves/myfavdocs/blob/main/make_files.sh#L114-L258) but is isn't rendering as expected when I add it to a docstring (https://raybellwaves.github.io/myfavdocs/generated/myfavdocs.pd_get_random_ingredients.html). Do I need to adjust the extension?

I saw something similar in the cudf project (https://docs.rapids.ai/api/cudf/stable/user_guide/api_docs/api/cudf.dataframe.transpose/#cudf.DataFrame.transpose https://github.com/rapidsai/cudf/blob/branch-24.06/docs/cudf/source/_ext/PandasCompat.py) as was curious if they had something different that makes this kind of extension work in a docstring

EDIT:

may be related to https://github.com/sphinx-doc/sphinx/issues/1837

**Comment by AA-Turner** [comment]
Hi Ray,

Are you able to provide a self-contained reproducer? Ideally a single conf.py file, index.rst, and a sample module for autodoc.

A

**Reply by raybellwaves** [reply]
https://github.com/raybellwaves/mcve-sphinxexttodo should be a minimal example. Apologies if it's missing something. Let me know and i'll be happy to work on it some more.

**Reply by raybellwaves** [reply]
<img width="1016" alt="Screenshot 2024-04-18 at 11 34 57 PM" src="https://github.com/sphinx-doc/sphinx/assets/17162724/abf6cf42-f19f-4adb-8922-8dad6f1d7580">

Looks ok in the default config so could be related to my config above

**Reply by raybellwaves** [reply]
Just making some notes as I investigate why it may not be working for my config.

I just tested with pydata-sphinx-theme and it still works as expected.

![Screenshot 2024-04-19 at 11 16 03 AM](https://github.com/sphinx-doc/sphinx/assets/17162724/fcf3e7de-cac0-4a88-a37a-21fe40266549)

**Reply by raybellwaves** [reply]
I see it now using sphinx.ext.napoleon and having a doc string like below. It wants to add the .. todo:: as another section in Returns

    """
    Return a list of random ingredients as strings.

    Parameters
    ----------
    kind : list[str]
        Optional “kind” of ingredients.

    Returns
    -------
    The ingredients list: list[str]

    .. todo:: This is in the docstring
    """

The site is hosted here https://raybellwaves.github.io/sphinxexttesting/, the code is here https://github.com/raybellwaves/sphinxexttesting and here's a screenshot:

![Screenshot 2024-04-19 at 11 26 45 AM](https://github.com/sphinx-doc/sphinx/assets/17162724/25fe061d-2441-44ca-a7bd-3899dfe13c6d)

**Reply by picnixz** [reply, marked as answer]
Ok I managed to make it work:

1. You need to include `todo_include_todos = True` (but you probably had it already).
2. You need *two* new lines before the `.. todo` directive, namely:

```python
"""
Return a list of random ingredients as strings.

Parameters
----------
kind : list[str]
    Optional “kind” of ingredients.

Returns
-------
The ingredients list: list[str]


.. todo:: ???
"""
```

![image](https://github.com/sphinx-doc/sphinx/assets/10796600/14b35439-0f39-45cc-bc37-6bc84142d7fb)

By the way, thanks to you I was able to find a totally unrelated bug which, when you put 4 "?", it doesn't work!
